app/assets/javascripts/angular/controllers/graph_ctrl.js

```diff
@@ -143,7 +147,32 @@ angular.module("Prometheus.controllers").controller('GraphCtrl',
     $scope.addAxis();
   }
 
+  $scope.$watch('graph.range', function() {
+    $scope.refreshGraph();
+  });
+  $scope.$watch('graph.endTime', function() {
+    $scope.refreshGraph();
+  });
+  $scope.$watch('graph.resolution', function() {
+    $scope.refreshGraph();
+  });
+
+  $scope.$watch(function() {
+    return $scope.graph.expressions.filter(function(e) {
+      // Filters newly added expressions from $watch, so it
```

Comment:
Wouldn't it be simpler and more consistent to also just emit a `setExpression` (probably then renamed to `changeExpression` to cover both changes to the text and the server) event for changes to the server ID? Then the code here can disappear and you'd only have to bind a change handler in the `expression.js.erb`. Together with ignoring empty expressions in general, that should be fine, or am I missing something?